Website Cart Should Not Show Expired Items As 'In Stock' #12535

tundebabzy opened the pull request:
Reference #12535 

Initially, the website used to include expired items. This PR fixes that.

![screenshot-2018-1-27 diced pork - 00002](https://user-images.githubusercontent.com/818803/35471889-a3d6f354-0364-11e8-8d9f-05eac988ab8e.png)

### Before
![screenshot-2018-1-27 diced pork](https://user-images.githubusercontent.com/818803/35471859-03562058-0364-11e8-9f15-e6404719f67d.png)

![screenshot-2018-1-27 products](https://user-images.githubusercontent.com/818803/35471862-06031680-0364-11e8-802d-7bcdc37f2a49.png)

### After
![screenshot-2018-1-27 diced pork 1](https://user-images.githubusercontent.com/818803/35471864-121ef056-0364-11e8-9c06-3a42f4898e7f.png)

![screenshot-2018-1-27 products 1](https://user-images.githubusercontent.com/818803/35471865-14e0893a-0364-11e8-8789-579e108492b7.png)

tundebabzy committed:
adjust stock_qty for expired quantities

adjust based on warehouse

add parameters to `get_qty_in_stock` so it can be useful in other parts of the code base

tundebabzy committed:
after adjusting stock_qty for expired, set in_stock flag

saurabh6790 commented on erpnext/utilities/product.py:
```diff
@@ -4,25 +4,70 @@
 	return frappe._dict({"in_stock": in_stock, "stock_qty": stock_qty, "is_stock_item": is_stock_item})
 
+
+def adjust_for_expired_items(item_code, stock_qty, warehouse):
```
`adjust_qty_for_expired_items`  maybe more specific ?

tundebabzy replied:
yeah

saurabh6790 commented on erpnext/utilities/product.py:
```diff
@@ -4,25 +4,70 @@
+	expired_batches = get_expired_batches(batches)
+	stock_qty = [list(item) for item in stock_qty]
+	
+	if expired_batches:
```
I think no need to check `if expired_batches` as `get_expired_batches` will return the empty list if expired batch not found

saurabh6790 commented on erpnext/setup/doctype/item_group/item_group.py:
```diff
@@ -104,8 +106,26 @@ def get_product_list_for_group(product_group=None, start=0, limit=10, search=Non
 	return [get_item_for_list_in_html(r) for r in data]
 
+
+def adjust_for_expired_items(data):
```
`adjust_qty_for_expired_items` maybe more specific ?

tundebabzy committed:
refactor `adjust_for_expired_items` and others as per code review

use get_all instead of get_list

rename `adjust_for_expired_items` to `adjust_qty_for_expired_items`

tundebabzy commented:
@saurabh6790 
fixed

saurabh6790 commented:
@tundebabzy Also remove `frappe.has_permission('Batch', throw=True)` from batch.py

tundebabzy committed:
remove stray code

tundebabzy commented:
@saurabh6790 done

saurabh6790 commented on erpnext/utilities/product.py:
```diff
@@ -4,25 +4,70 @@
 			item_code=%s and warehouse=%s""", (item_code, warehouse))
-		if stock_qty:
+
+	if stock_qty[0][0]:
```
If warehouse is not found @L21 then stock_qty will remain 0 and flow will break for stock_qty[0][0]

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object has no attribute '__getitem__'
```

so just check for stock_qty instead of stock_qty[0][0]

tundebabzy committed:
Merge branch 'hotfix' into issue-12535-1

tundebabzy committed:
code review fix

tundebabzy commented:
@saurabh6790 
Fixed
